android-agent/src/main/java/io/opentelemetry/android/internal/services/visiblescreen/VisibleScreenService.kt

```diff
@@ -0,0 +1,125 @@
+
+/**
+ * Wherein we do our best to figure out what "screen" is visible and what was the previously visible
+ * "screen".
+ *
+ * In general, we favor using the last fragment that was resumed, but fall back to the last
+ * resumed activity in case we don't have a fragment.
+ *
+ * We always ignore NavHostFragment instances since they aren't ever visible to the user.
+ *
+ * We have to treat DialogFragments slightly differently since they don't replace the launching
+ * screen, and the launching screen never leaves visibility.
+ */
+class VisibleScreenService internal constructor(private val application: Application) : Startable {
+    private val lastResumedActivity = AtomicReference<String>()
+    private val previouslyLastResumedActivity = AtomicReference<String>()
+    private val lastResumedFragment = AtomicReference<String>()
+    private val previouslyLastResumedFragment = AtomicReference<String?>()
+    private val activityLifecycleTracker by lazy { buildActivitiesTracker() }
+    private val fragmentLifecycleTrackerRegisterer by lazy { buildFragmentsTrackerRegisterer() }
+
+    companion object {
+        @JvmStatic
+        fun create(application: Application): VisibleScreenService {
+            return VisibleScreenService(application)
+        }
+    }
+
+    override fun start() {
+        application.registerActivityLifecycleCallbacks(activityLifecycleTracker)
+        application.registerActivityLifecycleCallbacks(fragmentLifecycleTrackerRegisterer)
+    }
+
+    private fun buildActivitiesTracker(): Application.ActivityLifecycleCallbacks {
+        return if (Build.VERSION.SDK_INT < 29) {
+            Pre29VisibleScreenLifecycleBinding(this)
+        } else {
+            VisibleScreenLifecycleBinding(this)
+        }
+    }
+
+    private fun buildFragmentsTrackerRegisterer(): Application.ActivityLifecycleCallbacks {
+        val fragmentLifecycle = VisibleFragmentTracker(this)
+        return if (Build.VERSION.SDK_INT < 29) {
+            RumFragmentActivityRegisterer.createPre29(fragmentLifecycle)
+        } else {
+            RumFragmentActivityRegisterer.create(fragmentLifecycle)
+        }
+    }
+
+    val previouslyVisibleScreen: String?
+        get() {
+            val previouslyLastFragment = previouslyLastResumedFragment.get()
+            if (previouslyLastFragment != null) {
+                return previouslyLastFragment
+            }
+            return previouslyLastResumedActivity.get()
+        }
+
+    val currentlyVisibleScreen: String
+        get() {
+            val lastFragment = lastResumedFragment.get()
+            if (lastFragment != null) {
+                return lastFragment
+            }
+            val lastActivity = lastResumedActivity.get()
+            if (lastActivity != null) {
+                return lastActivity
+            }
+            return "unknown"
+        }
+
+    fun activityResumed(activity: Activity) {
+        lastResumedActivity.set(activity.javaClass.simpleName)
```

Comment:
Do we have a R8 rule that or at least documentation that states activity class names shouldn't be obfuscated if users want the actual name of the activity to be used?

Reply:
Activity's are never minified, its part of the default rules

Reply:
(Does some reading)

Right, activities are referenced in the manifest so they can't be obfuscated. But fragment class names are obfuscated, and we try to do a similar thing too. I guess it doesn't matter so much if no one uses fragments anymore?

Reply:
I think it does matter, many apps still use fragments, or at least have a mixed setup.
IIRC we've discussed that before, ideally we'd offer a callback where the user can override the fragment's name based on the fragment instance/type, and if nothing is provided, we default to `simpleName` (which can be minified).

Reply:
I'm also learning more details about the existing instrumentations with these changes, so I think it's a great opportunity to bring these kinds of questions up. I've created [this issue](https://github.com/open-telemetry/opentelemetry-android/issues/459) to not loose track of this one, and added a couple of topics for tomorrow's SIG meeting.

Reply:
Another option is to add a proguard rule to ensure they aren't obfuscated. As long as we document that, it seems reasonable.